userspace/falco/falco_metrics.h

```diff
@@ -0,0 +1,40 @@
+// SPDX-License-Identifier: Apache-2.0
+/*
+Copyright (C) 2024 The Falco Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+#pragma once
+
+#include "configuration.h"
+
+#include <libsinsp/sinsp.h>
+
+namespace falco {
+	namespace app {
+		struct state;
+	}
+}
+
+class falco_metrics
+{
+public:
+	falco_metrics(falco::app::state& state);
+	bool is_enabled() const { return m_is_enabled; };
+	std::string to_text() const;
+
+private:
+	bool m_is_enabled = false;
+	std::vector<sinsp*> m_inspectors;
```

Comment:
@sgaist would you mind getting me up to speed why we need a vector with inspectors? Thanks!
Plus you still have an open question around sinsp and config init sequences etc?

Reply:
This is related to the initialization state of the sinsp objects when starting the webserver.

In my original implementation, all static stuff was created in the constructor of the `falco_metrics` but since the `machine_info` and `agent_info` objects are null pointers at that point, it's not doable. Hence I am currently keeping them as they cannot be retrieved through the converter class.

It's a temporary design to have something working that we can discuss about.

Reply:
Wonder if we can just adjust the order of initialization a bit? I can check later.

Reply:
@FedeDP @Andreagit97 plz advice, do we really need the vectors here because of the various source_infos. https://github.com/falcosecurity/falco/pull/3140/files#diff-0ebdd484eb8f6b10eb0df365c052ec7a91640bd6502ac3a7af1da82d9e62f3b1R31. We don't distinguish between the source info for the stats_writer either.

Reply:
I'd cc @jasondellaluce on this point, he might be able to help us :)

Reply:
Can't `falco_metrics::to_text` just take `falco::app::state&` as an argument? That way we avoid maintaining raw pointers all around the place. Also, why is `falco::app::state` declared with a forward declaration few lines above. What's the intention there?

Reply:
That's an interesting question/suggestion.

Just to make things clear for future readers, static, in this context, is meant as non-changing and is not related to the `static` C++ keyword.

My original design idea was the following:
1) pass the state to the constructor
2) build a set of static metrics (start up timestamps, etc.)
3) build a set of static textual metrics (falco version, kernel version etc.)
4) build a set of collectors for the dynamic values
5) build the final text in `to_text`

For 2 and 3, while both are static, they use different means to build the metrics text.

The goals for that design were:
1) generate metrics object for values that will not change only once
2) reduce the number of object creation/destruction cycle when requesting metrics (which is a direct follow up of 1).
3) allow to later change the collector class without having to rewrite the whole logic as it is basically: go through both sets of static metrics, go through dynamic metrics and voila, your metrics text is built.

This would make the metrics test generation faster although it would use a bit more memory to keep the static metrics values.

However, as noted earlier, for some reason I don't know (I currently don't know the libs code that well), it seems that the sinsp objects are not fully initialized when reaching the constructor otherwise, I would not have kept this vector of sinsp pointers.

As for the forward declaration, in this case, it's mainly out of habit from other projects I contribute to (only include where you use). Hence I am surprised I didn't do that for `sinsp` as well.

Note that I don't mind changing the design as it will mainly move the collectors creation to `to_text`.

Reply:
@FedeDP @jasondellaluce just a small ping around this design question.

I don't have any issue with re-implementing this part passing the application state as parameter of the `to_text` method, I would just like to have an agreement about which architecture we would like to adopt.

Reply:
@FedeDP @jasondellaluce @leogr +1 would appreciate final guidance on what you prefer so that we can get this over the finish line. Thanks.

Reply:
I don't have strong opinions, but I agree with this :point_down: 
> avoid maintaining raw pointers all around the place

Reply:
Re. the raw pointers, generally non-owning raw pointers are fine but the problems we can have is when these objects are free'd or reallocated by their owners.

I don't completely remember the complete mechanism that happens in the restart case, but will this work if Falco is restarted, e.g. when a rule is replaced and Falco auto-reloads (therefore rebuilding inspectors)? iirc it should but could you please double check?

Reply:
> I don't have any issue with re-implementing this part passing the application state as parameter of the to_text method, I would just like to have an agreement about which architecture we would like to adopt.

I see your point about creating non-mutating info at construction time, however I think that these non-owning raw pointers can create more harm than benefit. The issue is that they rely on leaked knowledge about:
- The order in which inspectors are allocated in the state, which is transitively dependent on how event sources are ordered, taking also for granted that this order won't change (which is like so now, but who knows later on)
- The lifecycle of these inspectors, meaning at "which point" in their initialization process they are (plugins, configs, captures being open, etc..), which is strongly dependent on the execution flow and the app actions ordering

So IMO it's not just the raw pointer (that could be solved by making it owner through a shared_ptr instance), but more that we leak implicit knowledge about other parts of the codebase. IMO, moving everything in `to_text` is clearer, meaning that the invoker of that method has responsibility of having everything in the right preconditions before invocation time, and doesn't have to care about small parts being retained. If re-computation is a concern, you can consider stateful lazy caching mechanisms (e.g. after getting a non-mutating metric the first time, save it in the state and don't re-compute it again).

I'm not gonna fight for this of course, but this is my point overall.

Reply:
In addition to your excellent points, I did not take into account the dynamic reloading use case which also goes in the favor of your suggested implementation so I will do that.

I think the lazy caching is a good idea for a follow up PR.